crypto/o_str.c

```diff
@@ -223,7 +223,18 @@ int openssl_strerror_r(int errnum, char *buf, size_t buflen)
 #if defined(_MSC_VER) && _MSC_VER>=1400
     return !strerror_s(buf, buflen, errnum);
 #elif defined(_GNU_SOURCE)
-    return strerror_r(errnum, buf, buflen) != NULL;
+    /* GNU strerror_r may not actually set buf.
+     * It may return a pointer to some (immutable) static string in which case
+     * buf is left unused.
+     */
+    char *err = strerror_r(errnum, buf, buflen);
+    if (err == NULL)
+        return 0;
+    if (!*buf) {
```

Comment:
AFAICT from the `strerror_r` man page, `buf` is left completely untouched in the event that it returns a pointer to some static immutable string. In which case (presumably) `*buf` contains whatever it contained on entry to this function - which may or may not be == 0. Should we be setting `*buf = '\0'` before we make the call to `strerror_r`?

Also style nit: please write `!*buf` as `*buf == '\0'`

Reply:
That's correct, I didn't check the other openssl_strerror_r calls.
The buf is definitely not zeroed in dso_dl.c and loader_file.c.

What about a different approach?

`if (*buf == '\0)`
could be replaced by
`if (strncmp(err, buf, buflen) != 0)`

There won't be a need for changing the usage of openssl_strerror_r.

Reply:
Yes, that would work too (although not particularly efficient). In the event that the buffer is filled is it the case that `err == buff`, or is that not guaranteed?

> There won't be a need for changing the usage of openssl_strerror_r.

If you set `*buf = '\0'` before the strerror_r call then there still wouldn't be a need to change the usage of openssl_strerror_r (although it would be a little odd that `buf` would get modified even in the eventuality that strerror_r fails (returns NULL)).

Reply:
> Yes, that would work too (although not particularly efficient). In the event that the buffer is filled is it the case that `err == buf`, or is that not guaranteed?

Looking at the man page, it's not guaranteed.
The error string could be truncated to fit buf should it be longer than buflen.

> If you set `*buf = '\0'` before the strerror_r call then there still wouldn't be a need to change the usage of openssl_strerror_r (although it would be a little odd that `buf` would get modified even in the eventuality that strerror_r fails (returns NULL)).

I kind of like this idea the most.